modules/aruco/include/opencv2/aruco.hpp

```diff
@@ -0,0 +1,502 @@
+ * are provided, (e.g std::vector<std::vector<cv::Point2f> > ). For N detected markers,
+ * the dimensions of this array should be Nx4. The order of the corners should be clockwise.
+ * @sa detectMarkers
+ * @param markerLength the length of the markers' side. The returning translation vectors will
+ * be in the same unit. Normally, unit is meters.
+ * @param cameraMatrix input 3x3 floating-point camera matrix
+ * \f$A = \vecthreethree{f_x}{0}{c_x}{0}{f_y}{c_y}{0}{0}{1}\f$
+ * @param distCoeffs vector of distortion coefficients
+ * \f$(k_1, k_2, p_1, p_2[, k_3[, k_4, k_5, k_6],[s_1, s_2, s_3, s_4]])\f$ of 4, 5, 8 or 12 elements
+ * @param rvecs array of output rotation vectors (@sa Rodrigues) (e.g. std::vector<cv::Mat>>).
+ * Each element in rvecs corresponds to the specific marker in imgPoints.
+ * @param tvecs array of output translation vectors (e.g. std::vector<cv::Mat>>).
+ * Each element in tvecs corresponds to the specific marker in imgPoints.
+ *
+ * This function receives the detected markers and returns their pose estimation respect to
+ * the camera individually. So for each marker, one rotation and translation vector is returned.
+ * The returned transformation is the one that transforms points from each marker coordinate system
+ * to the camera coordinate system.
+ * The marker corrdinate system is centered on the middle of the marker, with the Z axis
+ * perpendicular to the marker plane.
+ * The coordinates of the four corners of the marker in its own coordinate system are:
+ * (-markerLength/2, markerLength/2, 0), (markerLength/2, markerLength/2, 0),
+ * (markerLength/2, -markerLength/2, 0), (-markerLength/2, -markerLength/2, 0)
+ */
+CV_EXPORTS void estimatePoseSingleMarkers(InputArrayOfArrays corners, float markerLength,
+                                          InputArray cameraMatrix, InputArray distCoeffs,
+                                          OutputArrayOfArrays rvecs, OutputArrayOfArrays tvecs);
+
+
+
+/**
+ * @brief Board of markers
+ *
+ * A board is a set of markers in the 3D space with a common cordinate system.
+ * The common form of a board of marker is a planar (2D) board, however any 3D layout can be used.
+ * A Board object is composed by:
+ * - The object points of the marker corners, i.e. their coordinates respect to the board system.
+ * - The dictionary which indicates the type of markers of the board
+ * - The identifier of all the markers in the board.
+ */
+class CV_EXPORTS Board {
+
+    public:
+    // array of object points of all the marker corners in the board
+    // each marker include its 4 corners, i.e. for M markers, the size is Mx4
+    std::vector< std::vector< Point3f > > objPoints;
+
+    // the dictionary of markers employed for this board
+    Dictionary dictionary;
+
+    // vector of the identifiers of the markers in the board (same size than objPoints)
+    // The identifiers refers to the board dictionary
+    std::vector< int > ids;
```

Comment:
I really like the simple design of Board and related data. But, I think Board itself could be an ```std::map``` or ```std::unordered_map``` with key being id and value being ```objPoints[index_corresponding_to_key]``` to allow for fast query of 3D objectPoints. Did you have a particular reason in terms of memory efficiency to choose vectors? We could measure how memory performance is when Charuco runs.